src/ln/channelmonitor.rs

```diff
@@ -516,38 +603,82 @@ impl ChannelMonitor {
 			feerate_per_kw,
 			htlc_outputs,
 		});
-		self.key_storage = if let KeyStorage::PrivMode { ref revocation_base_key, ref htlc_base_key, ref delayed_payment_base_key, ref payment_base_key, ref shutdown_pubkey, ref latest_per_commitment_point, .. } = self.key_storage {
-			KeyStorage::PrivMode {
+		self.key_storage = if let Storage::Local { ref revocation_base_key, ref htlc_base_key, ref delayed_payment_base_key, ref payment_base_key, ref shutdown_pubkey, ref latest_per_commitment_point, ref funding_info, ref short_channel_id, .. } = self.key_storage {
+			Storage::Local {
 				revocation_base_key: *revocation_base_key,
 				htlc_base_key: *htlc_base_key,
 				delayed_payment_base_key: *delayed_payment_base_key,
 				payment_base_key: *payment_base_key,
 				shutdown_pubkey: *shutdown_pubkey,
 				prev_latest_per_commitment_point: *latest_per_commitment_point,
 				latest_per_commitment_point: Some(local_keys.per_commitment_point),
+				funding_info: funding_info.clone(),
+				short_channel_id: *short_channel_id,
 			}
 		} else { unimplemented!(); };
 	}
 
 	/// Provides a payment_hash->payment_preimage mapping. Will be automatically pruned when all
 	/// commitment_tx_infos which contain the payment hash have been revoked.
-	pub(super) fn provide_payment_preimage(&mut self, payment_hash: &[u8; 32], payment_preimage: &[u8; 32]) {
+	pub(super) fn provide_payment_preimage(&mut self, payment_hash: &PaymentHash, payment_preimage: &PaymentPreimage) {
 		self.payment_preimages.insert(payment_hash.clone(), payment_preimage.clone());
 	}
 
 	/// Combines this ChannelMonitor with the information contained in the other ChannelMonitor.
 	/// After a successful call this ChannelMonitor is up-to-date and is safe to use to monitor the
 	/// chain for new blocks/transactions.
 	pub fn insert_combine(&mut self, mut other: ChannelMonitor) -> Result<(), MonitorUpdateError> {
-		if self.funding_txo.is_some() {
-			// We should be able to compare the entire funding_txo, but in fuzztarget its trivially
-			// easy to collide the funding_txo hash and have a different scriptPubKey.
-			if other.funding_txo.is_some() && other.funding_txo.as_ref().unwrap().0 != self.funding_txo.as_ref().unwrap().0 {
-				return Err(MonitorUpdateError("Funding transaction outputs are not identical!"));
-			}
-		} else {
-			self.funding_txo = other.funding_txo.take();
-		}
+
+		self.key_storage = match self.key_storage {
```

Comment:
I really dont think we need to support all these cases - at no point should we get called on a local monitor that doesn't have funding_info set, so that can just return an Err (which means the comparison can just go ahead as it was previously). Also, we should never get a call to merge a watchtower-stripped monitor with a local monitor, so that can just Err.

Reply:
Hey simplify on this one, but maybe not as much as you was thinking to avoid breaking some tests